test/performance/odp_l2fwd.c

```diff
@@ -434,6 +434,18 @@ static int run_worker_sched_mode(void *arg)
 		stats->s.packets += pkts;
 	}
 
+	/* Free prefetched packets */
+	odp_schedule_pause();
+	while (1) {
+		odp_event_t  ev;
+
+		ev = odp_schedule(NULL, ODP_SCHED_NO_WAIT);
+		if (ev == ODP_EVENT_INVALID)
+			break;
+		odp_event_free(ev);
+	}
+	odp_schedule_resume();
```

Comment:
At this point pktios have not necessarily been stopped yet, so new events can be available and maybe the scheduler could do some prefetching even without an explicit schedule call by this thread (e.g. HW scheduler prefetching event asynchronously or SW scheduler or SW packet reception in some other thread pushing packets to this thread. Then we would be in the same state as before the pause-resume dance.

ODP API guarantees that stale locally prefetched events will not be left dangling is one first calls pause and then drains the scheduler locally by calling schedule. I would not like to depend on the exact behavior when not following that sequence. If the resume call were moved little bit forward so that it comes after the barrier wait, then this issues
would seem to go away as at that point the pktios have already been stopped and all threads have come out of the forwarding loop.

Reply:
Moving resume call below barrier makes sense. Will update the patch.